exp/runtime/hooks/api/v1alpha1/discovery_types.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"sigs.k8s.io/cluster-api/internal/runtime/catalog"
+)
+
+// FailurePolicy is the type of a extension failure policy.
+type FailurePolicy string
+
+const (
+	// FailurePolicyIgnore means that an error calling the extension is ignored.
+	FailurePolicyIgnore FailurePolicy = "Ignore"
+
+	// FailurePolicyFail means that an error calling the extension causes the admission to fail.
+	FailurePolicyFail FailurePolicy = "Fail"
+)
+
+// Hook is a identifier of a runtime hook.
+type Hook struct {
+	// APIVersion is the Version of the Hook
+	APIVersion string `json:"apiVersion"`
+
+	// Name is the name of the hook
+	Name string `json:"name"`
+}
+
+// RuntimeExtension represents the discovery information of the extension which includes
+// the hook is supports.
+type RuntimeExtension struct {
+	// Name is the name of the RuntimeExtension
+	Name string `json:"name"`
+
+	// Hook defines the specific runtime event for which this RuntimeExtension calls.
+	Hook Hook `json:"hook"`
+
+	// TimeoutSeconds defines the timeout duration for client calls to the Hook
+	TimeoutSeconds *int32 `json:"timeoutSeconds,omitempty"`
+
+	// FailurePolicy defines how failures in calls to the Hook should be handled by a client.
+	FailurePolicy *FailurePolicy `json:"failurePolicy,omitempty"`
+}
+
+// DiscoveryHookRequest foo bar baz.
```

Comment:
```suggestion
// DiscoveryHookRequest is a Discovery request.
```
or something 

Should we drop `Hook`? => DiscoveryRequest + DiscoveryReponse